File: RemEx.Android/app/src/main/java/com/clindsay94/remex/security/TransportTrust.kt
package com.clindsay94.remex.security

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities

/**
 * Classifies the network path to a host so the pairing flow can decide whether the
 * channel is already authenticated and MITM-resistant end-to-end — and therefore
 * trustworthy enough to auto-fetch the 6-digit pairing PIN over it — or an untrusted
 * LAN/internet path where the PIN must keep its purpose as a genuine out-of-band
 * secret and be entered manually.
 *
 * "Trusted transport" means: loopback (same device) or a Tailscale/WireGuard tunnel
 * (mutually authenticated, encrypted, no on-path attacker). On any other transport the
 * PIN is the only thing standing between an active network attacker and a successful
 * first-pair, so it must NOT be served to the client automatically.
 */
object TransportTrust {

    fun isLoopback(host: String): Boolean {
        val h = host.trim().lowercase()
        return h == "localhost" || h == "::1" || h.startsWith("127.")
    }

    /**
     * Tailscale assigns addresses from the CGNAT range 100.64.0.0/10 (IPv4) and the
     * fd7a:115c:a1e0::/48 ULA prefix (IPv6). Such a host is only reachable through the
     * WireGuard tunnel, so being able to reach it at all implies the tunnel is up and
     * the remote peer has been authenticated by Tailscale.
     */
    fun isTailscaleAddress(host: String): Boolean {
        val h = host.trim().lowercase()
        if (h.startsWith("fd7a:115c:a1e0:")) return true
        val octets = h.split(".")
        if (octets.size != 4) return false
        val first = octets[0].toIntOrNull() ?: return false
        val second = octets[1].toIntOrNull() ?: return false
        return first == 100 && second in 64..127
    }

    /** True when the device currently routes through a VPN transport (e.g. Tailscale). */
    fun isVpnActive(context: Context): Boolean {
        return try {
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
                ?: return false
            val caps = cm.getNetworkCapabilities(cm.activeNetwork) ?: return false
            caps.hasTransport(NetworkCapabilities.TRANSPORT_VPN)
        } catch (_: Exception) {
            false
        }
    }

    /**
     * Whether the pairing PIN may be auto-fetched over the network to [host]. Allowed
     * only for loopback, or for a Tailscale address while a VPN tunnel is actually
     * active. Everything else requires manual, out-of-band PIN entry.
     */
    fun canAutoFetchPin(context: Context, host: String): Boolean {
        if (isLoopback(host)) return true
        return isTailscaleAddress(host) && isVpnActive(context)
    }

    /**
     * Whether connecting to [host] needs LAN-scoped runtime permissions
     * (NEARBY_WIFI_DEVICES / ACCESS_LOCAL_NETWORK). A loopback or VPN/Tailscale target
     * is not on the local network, so those permissions are irrelevant and must never
     * block the connection — that was the cause of "won't even try to connect" when a
     * user denied local-network access but reached the host over Tailscale.
     */
    fun requiresLocalNetworkAccess(host: String): Boolean {
        return !(isLoopback(host) || isTailscaleAddress(host))
    }
}

File: remex.agent/Services/Security/TransportTrust.cs
using System;
using System.Net;
using System.Net.Sockets;

namespace Remex.Agent.Services.Security;

/// <summary>
/// Classifies the network path a request arrived over so the pairing endpoints can decide whether
/// the channel is already mutually-authenticated and MITM-resistant — and therefore trustworthy
/// enough to relay the 6-digit pairing PIN automatically — or an untrusted LAN/internet path where
/// the PIN must keep its purpose as a genuine out-of-band secret and be entered by hand.
/// </summary>
/// <remarks>
/// This is the host-side mirror of the Android client's <c>TransportTrust</c>
/// (<c>remex.android/.../security/TransportTrust.kt</c>); both sides must agree on what counts as a
/// "trusted transport" or the PIN auto-fill silently fails on one end. "Trusted" means loopback
/// (same machine) or an active Tailscale/WireGuard tunnel. Tailscale assigns addresses from the
/// CGNAT range <c>100.64.0.0/10</c> (IPv4) and the <c>fd7a:115c:a1e0::/48</c> ULA prefix (IPv6);
/// such an address is only reachable through the WireGuard tunnel, so a request arriving on one
/// implies the tunnel is up and the peer has been authenticated by Tailscale.
/// Pure and static so it is unit-testable without a live connection.
/// </remarks>
public static class TransportTrust
{
    // First 6 bytes of the fd7a:115c:a1e0::/48 Tailscale IPv6 ULA prefix.
    private static readonly byte[] TailscaleV6Prefix = { 0xfd, 0x7a, 0x11, 0x5c, 0xa1, 0xe0 };

    /// <summary>True when <paramref name="ip"/> is an IPv4/IPv6 loopback address (same machine).</summary>
    public static bool IsLoopback(IPAddress? ip) => ip is not null && IPAddress.IsLoopback(ip);

    /// <summary>
    /// True when <paramref name="ip"/> falls in a Tailscale-assigned range: IPv4 <c>100.64.0.0/10</c>
    /// or IPv6 <c>fd7a:115c:a1e0::/48</c>. IPv4-mapped IPv6 addresses are unwrapped first.
    /// </summary>
    public static bool IsTailscaleAddress(IPAddress? ip)
    {
        if (ip is null)
        {
            return false;
        }

        // Kestrel may surface an IPv4 peer as an IPv4-mapped IPv6 address (::ffff:100.64.x.x).
        if (ip.IsIPv4MappedToIPv6)
        {
            ip = ip.MapToIPv4();
        }

        if (ip.AddressFamily == AddressFamily.InterNetwork)
        {
            // 100.64.0.0/10: first octet 100, second octet 64..127.
            var octets = ip.GetAddressBytes();
            return octets[0] == 100 && octets[1] >= 64 && octets[1] <= 127;
        }

        if (ip.AddressFamily == AddressFamily.InterNetworkV6)
        {
            var bytes = ip.GetAddressBytes();
            for (var i = 0; i < TailscaleV6Prefix.Length; i++)
            {
                if (bytes[i] != TailscaleV6Prefix[i])
                {
                    return false;
                }
            }

            return true;
        }

        return false;
    }

    /// <summary>
    /// Whether the active pairing PIN may be served to a caller. Allowed only when the caller is on
    /// loopback, or when both the caller (<paramref name="remote"/>) and the host-side address the
    /// caller reached (<paramref name="local"/>) are Tailscale addresses — i.e. the connection genuinely
    /// traversed the Tailscale tunnel. Requiring <paramref name="local"/> to also be Tailscale defeats a
    /// LAN attacker who spoofs a <c>100.64.x.x</c> source while connecting to the host's LAN IP (there the
    /// host-side address is the LAN IP, not a Tailscale one). Everything else must enter the PIN manually.
    /// </summary>
    public static bool IsTrustedForPinAutoFetch(IPAddress? remote, IPAddress? local) =>
        IsLoopback(remote) || (IsTailscaleAddress(remote) && IsTailscaleAddress(local));
}
